### PR TITLE
CD-38: use Flexbox/Grid to make top-nav more robust

### PR DESCRIPTION
**Pull request by rupl**
What I thought must be some typo in the media queries ended up being three unrelated special cases in the code, one of them being the actual content of the nav. That's sure to cause problems on a CMS-powered navigation row so I went ahead and started trying to clear this up.

It's not the prettiest, but I didn't want to rewrite the five or so Sass partials that I had to touch to make this happen.. The PR gets rid of many floats and uses two systems to align everything:

* CSS Grid for the highest-level header arrangement. Logo + "Actions"
* "Actions" is a flexbox: Search + Nav
* Nav is also a Flexbox, forcing everything onto one line at the Desktop breakpoint and ensuring that long nav items don't push the last ones off the row. They wrap in their containers instead, which is ugly but arguably more orderly than nav items sitting next page content.

It leaves all the old code in place for IE11 and wraps the enhancements in a compound Modernizr class that requires both features to be fully supported before the layout kicks in: `.cssgrid.flexbox`

### CODE REVIEW

File: common-design/sass/cd-header/_cd-header.scss [hunk 1]
```diff
@@ -29,8 +29,41 @@
 
 .cd-site-header__inner {
   height: $cd-site-header-height;
+  display: grid;
+  grid-template-columns: 52px 1fr;
+  grid-template-rows: $cd-site-header-height;
+
+  @media (min-width: 768px) {
+    grid-template-columns: 186px 1fr;
+  }
```
Comment (rupl) [inline common-design/sass/cd-header/_cd-header.scss]: reminder to myself to swap out hardcoded MQ with the current convention in the codebase

File: common-design/sass/cd-header/_cd-header.scss [hunk 2]
```diff
@@ -29,8 +29,41 @@
...
+}
+
+.cd-nav > ul {
+  display: flex;
+  flex-flow: row nowrap;
+  justify-content: flex-end;
+
+  & > .cd-nav__item {
+    flex: 1 0 14.286%;
+  }
```
Comment (rupl) [inline common-design/sass/cd-header/_cd-header.scss]: 🤢 at this set of selectors but Flexbox only works on direct descendants.. the `<ul>` is in the way unless it's considered the flex container, plus any deeper children need to be ignored.

**Comment by rupl** [comment]
### Before

Nav is dropping below nav area into Content, causing content to flow around it.

menu-before

### After

menu-after

**Comment by rupl** [comment]
@left23 this is now ready for your review. It's friday so feel free to kick it to next week 😁

File: common-design/css/styles.css [hunk 3]
```diff
@@ -427,7 +475,6 @@ ul.cd-nav__grandchild-nav {
       display: -webkit-flex;
       display: flex;
       height: 60px;
-      line-height: 60px;
```
Comment (rupl) [inline common-design/css/styles.css:430]: this line didn't add anything, and it messed up the squished nav links when they were forced to wrap as flex-items.

**Review by left23 — DISMISSED** [review]
@rupl Overall I think this is a good change. Thanks!

I think adding some space to the right of the logo will help its appearance (and match it closer to the original design) and also better facilitate the inline-search (which has a left-border separating it from the logo).

The menu items are left align which I assume is to facilitate the cases where the menu item text wraps. I'm not too concerned about this (visually it looks fine to me) but I wonder about the designer's opinion. I guess the more items, the less noticeable.

**Comment by rupl** [comment]
Good point yes that was a mistake on my part. I'll make it look like the original.

**Comment by rupl** [comment]
yeah I totally overthought this. Now the code lets flexbox do what flexbox does.

**Comment by rupl** [comment]
Here's a GIF showing parity with existing layout:

cd-38-flex-nav

**Comment by emmajane** [comment]
nnneaaaaar faaarrrrrr

**Comment by rupl** [comment]
Hm, if the nav has extra items it still overflows with logo. Example:

cd-38-overflow

**Comment by rupl** [comment]
ok maybe this is it? kinda going cross-eyed looking at this so hopefully it makes sense when I sit back down tomorrow. My gh-pages is updated: https://rupl.github.io/styleguide/common-design/demo.html

_edit:_ I also included longer nav items in my staging site. I pulled the nav from a real site and even added a word or two to make it long as possible without exaggerating

**Comment by left23** [comment]
This looks good to me. I understand it's fiddly but I consider it a nice improvement and allows more flexibility in general

**Comment by rupl** [comment]
alright! I discovered there's a breakpoint beyond `desktop` that I didn't know about.. that allows for wider nav on big screens without messing up all the breakpoints below. Now I'm ok with it. :shipit:

**Pull request merged**